### PR TITLE
[WIP] use inline config or config file for oc

### PR DESCRIPTION
**Pull request by rinormaloku**
This resolves issue #146 and adds the possibility to use **your own configuration file** and or **use inline configuration** (similar to kubectl task).

Kept it at WIP as if those changes are wanted I'd update the documentation accordingly

### CODE REVIEW

**Comment by lstocchi** [comment]
Hey @rinormaloku, thanks for your contribution. I gave a look at your PR and I found it interesting. I like the idea of giving users the opportunity to specify a path for the configuration file to be used. Especially when using a self-hosted agent.

What i don't like is the way user can set up the connection to the cluster. Currently users need to go to the web platform to set up a cluster connection (inside service connections). By giving the option to set up a connection from a pipeline task takes to a consistency error with how the extension currently works. My doubt is ... if we give the opportunity to write a kubectl config as an input why shouldn't we give the chance to also allow to add a token or username/password as an input and change the OpenShift connection created previously? But, by doing this it could create some security problems because users should start using secrets or their data could be visible to everyone.

Is there something i'm missing?

**Comment by rinormaloku** [comment]
Hi @lstocchi 

The users will be able to add the configuration as a secret, in the same approach that is used in Azure DevOps by using the $(SECRET) pattern wherever they want to replace a value.

As such this is no security issue (besides if someone wants to paste the configuration directly in the inline config). This feature is to me just a bonus to keep the same expectation as in Kubectl.

The way more important one is being able to specify the location of kubeconfig, for cases when using an agent.

To your suggestion that this can be extended to include cases with a token, I totally agree (though expiry of tokens could be a problem, and I'd have to read up on that)

**Comment by lstocchi** [comment]
@rinormaloku I tried using the Kubectl task but i couldn't find where they allow to set up a configuration at task level. As far as i saw, first you need to set up the config in the service connections panel and then they allow you to pick up the connection you want to use at task level which is quite similar to what we allow here.

image
image

I would suggest to keep everything in the service connections panel and edit/remove the no_authentication option (which i think it's confusing, we should also change the name) by giving the user the option to choose between an inline config or a configuration file path. This will also keep the number of inputs as it is now, which is good.

**Comment by rinormaloku** [comment]
@lstocchi here is how they achieve it:
image

I made some adjustments as it was not intuitive. 'None' option didn't make sense. And whatever option you choose you still had the possibility to check use configuration and select one of the options (File path or inline config). As such here is how I thought looks more intuitive:

1. Select one of the options "OpenShift Connection Service" is selected by default.
image

2. According to selection you get different options, e.g. file config:
image

This was my reasoning for making the changes. 

If I understood you correctly you'd prefer the same approach kubectl took?

**Comment by lstocchi** [comment]
@rinormaloku ah, thanks for clarify how they do this. I missed that part.

I still have many doubts. I don't like the way they manage this. As you said there is no reason to keep the 'use configuration' check when not required. It's not intuitive and confusing. 

What i need to think about is when it's useful for a user to change connection configuration at task level. Let's suppose i'm working with a different cluster (not the usual one) and i want to run my task on it, i would rather create a different service connection and specify which connection i want to use, compared on giving the user the possibility to set up the configuration on the task itself. It helps to keep the pipeline cleaner. 

My current thought would be to keep everything in the service connection panel so that user could create its service connection by using either the token or username/password or inline config or a config path.

I'll think at it.

**Comment by rinormaloku** [comment]
> What i need to think about is when it's useful for a user to change connection configuration at task level. Let's suppose i'm working with a different cluster (not the usual one) and i want to run my task on it, i would rather create a different service connection and specify which connection i want to use, compared on giving the user the possibility to set up the configuration on the task itself. It helps to keep the pipeline cleaner.

In my case, for example, I don't have access to the kubeconfig or credentials, those are being retrieved at runtime (from Azure Key Vault) and I need to specify the directory from which to read it.

**Comment by lstocchi** [comment]
> > What i need to think about is when it's useful for a user to change connection configuration at task level. Let's suppose i'm working with a different cluster (not the usual one) and i want to run my task on it, i would rather create a different service connection and specify which connection i want to use, compared on giving the user the possibility to set up the configuration on the task itself. It helps to keep the pipeline cleaner.
> 
> In my case, for example, I don't have access to the kubeconfig or credentials, those are being retrieved at runtime (from Azure Key Vault) and I need to specify the directory from which to read it.

AFAIK you can set secrets in Azure Key Vault and retrieve their value in Azure DevOps this way. First, you set up a service connection to connect to your Azure account and then you can create variables to contain secrets retrieved from Azure Key Vault and use them in a pipeline. Right?

But, if you set your secrets in your Azure Key Vault you know which is your kubeconfig or credentials. Otherwise how could you set the value of your secrets? So why preferring this approach (creating a secret, then a variable on DevOps, then insert them in the pipeline) when someone could simply create a service connection? 

The only thing i can think of is by maintaining all sensitive data in one place (Azure Key Vault) so if they need to change something they don't need to go over all service connections and edit them. Correct? If this is the case i would change the options we show to users. I would move everything in the backend (inside service connections) so people can still set up their connections by also using kubeconfig path or inline config + adding (at task level) an option to overwrite the config and make user pick up the variable (which contains the config retrieved by Azure Key Vault) and use that as config. Is that reasonable? I guess I still miss something, if you could enlighten me i would really appreciate it.

**Comment by rinormaloku** [comment]
@lstocchi in my case the config is generated in the earlier stages of the pipeline. But that's beside the point :P my case is kinda special, and others might need to specify the kubeconfig path for other cases.

Though I specifically do not need this update in the extension as I am using my version. If you think that this won't be needed for others then we can close it. In the other side if you think we can make changes to make it even more useful. I'm all for it :)

**Comment by lstocchi** [comment]
@rinormaloku yeah i find this useful and i want to add this to the extension (I can also edit your PR by myself if you prefer). What i want to understand better is the correlation between Azure Key Vault and the path of the kubeconfig. I know you can store a file in Azure Key Vault but when you try to retrieve it, you will get the content of the file stored (not the path), correct? 
What do you mean by "the config is generated at early stage"? Do you store the value retrieved from Azure Key Vault inside a variable and use it in your pipeline? Or you do this in another way (i'm not aware of)? If you could copy part of your pipeline would be great.

**Comment by rinormaloku** [comment]
@lstocchi at the risk of taking the conversation completely off-topic, here is why I do it:

I  setup an OCP 4.x cluster, and asap the cluster is up, I proceed to install some of the components which are important for me (e.g. Quay, EFK etc.) so my release pipeline is separated in those stages:
1. Setup Azure Prerequisites (Which creates the keyvault)
2. Setup OCP cluster (upload kubeconfig to keyvault)
3. Install resources in the created cluster (pull kubeconfig from keyvault and store to a file)

After pulling it to a file, if I want to use the extension and for this reason I need to specify the kubeconfig location.
 
Service Connections would require me to break the pipeline into multiple pipelines so that I can create the Service Connection and update the pipeline. Not really enjoyable.

**Comment by lstocchi** [comment]
Ah ok, so basically the problem is that azure still doesn't support a simple way to pass a variable through different stages (https://github.com/microsoft/azure-pipelines-tasks/issues/4743) and as a workaround you store the value retrieved from the Key Vault into a file and use it later on.

**Comment by rinormaloku** [comment]
I am not sure if DevOps allows passing on values, my main goal is to have a one-click deployment for the cluster and additional preinstalled components (quay, efk, ...).

OpenShift Service Connections requires a manual step (which I am happy to avoid).

File: tasks/config-map/task.json [hunk 1]
```diff
@@ -12,27 +12,81 @@
...
   "inputs": [
+    {
+      "name": "connectionType",
+      "type": "pickList",
+      "label": "Service connection type",
+      "defaultValue": "OpenShift Connection Service",
+      "required": true,
+      "options": {
+        "OpenShift Connection Service": "OpenShift Connection Service",
+        "File Configuration": "File Configuration",
```
Comment (lstocchi) [inline tasks/config-map/task.json]: I would follow the way kubectl works. I'd suggest substitute them with a new entry (e.g "Runtime Configuration": "Set Up Configuration on Runtime") and then when selected, show the radio type with the two options you already created (file config, inline config.).

File: tasks/oc-cmd/task.json [hunk 2]
```diff
@@ -12,27 +12,80 @@
...
   "inputs": [
+    {
+      "name": "connectionType",
+      "type": "pickList",
+      "label": "Service connection type",
+      "defaultValue": "OpenShift Connection Service",
+      "required": true,
+      "options": {
+        "OpenShift Connection Service": "OpenShift Connection Service",
+        "File Configuration": "File Configuration",
```
Comment (lstocchi) [inline tasks/oc-cmd/task.json]: Same as above

File: tasks/oc-setup/task.json [hunk 3]
```diff
@@ -12,27 +12,81 @@
...
   "inputs": [
+    {
+      "name": "connectionType",
+      "type": "pickList",
+      "label": "Service connection type",
+      "defaultValue": "OpenShift Connection Service",
+      "required": true,
+      "options": {
+        "OpenShift Connection Service": "OpenShift Connection Service",
+        "File Configuration": "File Configuration",
```
Comment (lstocchi) [inline tasks/oc-setup/task.json]: Same as above

**Comment by lstocchi** [comment]
I'll merge this and fix the remaining stuff in another PR because i'm refactoring the extension a bit. Thanks for your help.

**Pull request merged**